kartothek/api/consistency.py

```diff
@@ -0,0 +1,355 @@
+        if unspecified_partition_columns:
+            raise ValueError(
+                f"Unspecified but provided partition columns in {klee_dataset_id}: "
+                f"{', '.join(sorted(unspecified_partition_columns))}"
+            )
+
+
+def _check_indices(datasets, cube):
+    """
+    Check if required indices are present in given datasets.
+
+    For all datasets the primary indices must be equal to ``ds.partition_keys``. For the seed dataset secondary
+    indices for all dimension columns are expected.
+
+    Parameters
+    ----------
+    datasets: Dict[str, kartothek.DatasetMetadata]
+        Datasets.
+    cube: kartothek.core.cube.cube.Cube
+        Cube specification.
+
+    Raises
+    ------
+    ValueError: In case indices are broken.
+    """
+    for klee_dataset_id in sorted(datasets.keys()):
+        ds = datasets[klee_dataset_id]
+        primary_indices = ds.partition_keys
+        columns = get_dataset_columns(ds)
+        secondary_indices = set()
+        any_indices = set(cube.index_columns) & columns
+
+        if klee_dataset_id == cube.seed_dataset:
+            secondary_indices |= set(cube.dimension_columns)
+
+        for types, elements in (
+            ((PartitionIndex,), primary_indices),
+            ((ExplicitSecondaryIndex,), secondary_indices),
+            ((ExplicitSecondaryIndex, PartitionIndex), any_indices),
+        ):
+            tname = " or ".join(t.__name__ for t in types)
+
+            # it seems that partition indices are not always present (e.g. for empty datasets), so add partition keys to
+            # the set
+            indices = copy(ds.indices)
+            if PartitionIndex in types:
+                for pk in ds.partition_keys:
+                    if pk not in indices:
+                        indices[pk] = "dummy"
+
+            for e in sorted(elements):
+                if e not in indices:
+                    raise ValueError(
+                        '{tname} "{e}" is missing in dataset "{klee_dataset_id}".'.format(
+                            tname=tname, e=e, klee_dataset_id=klee_dataset_id
+                        )
+                    )
+
+                idx = indices[e]
+                t2 = type(idx)
+                tname2 = t2.__name__
+                if (idx != "dummy") and (not isinstance(idx, types)):
+                    raise ValueError(
+                        '"{e}" in dataset "{klee_dataset_id}" is of type {tname2} but should be {tname}.'.format(
+                            tname=tname,
+                            tname2=tname2,
+                            e=e,
+                            klee_dataset_id=klee_dataset_id,
+                        )
+                    )
+
+
+def check_datasets(datasets, cube):
+    """
+    Apply sanity checks to persisteted Karothek datasets.
+
+    The following checks will be applied:
+
+    - seed dataset present
+    - metadata version correct
+    - only the klee-specific table is present
```

Comment:
don't want to nitpick docs but now is the best time I guess. This is referring to a kartothek dataset-table not to a cube table, correct? We should probably be very explicit with the nomenclature everywhere

cc @marco-neumann-jdas

Reply:
changed the nomenclature to ktk_cube  or kartothek wherever applicable.